src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexer.cs

```diff
@@ -84,205 +103,340 @@ public void Initialize()
+                    if (this.cancellation.IsCancellationRequested)
+                        break;
 
-                this.logger.LogCritical(message);
-                this.logger.LogTrace("(-)[INCONSISTENT_STATE]");
-                throw new Exception(message);
-            }
+                    ChainedHeader nextHeader = this.consensusManager.Tip.GetAncestor(this.IndexerTip.Height + 1);
 
-        }
+                    if (nextHeader == null)
+                    {
+                        this.logger.LogDebug("Next header wasn't found. Waiting.");
 
-        private Task OnEnqueueAsync(KeyValuePair<bool, ChainedHeaderBlock> item, CancellationToken cancellationtoken)
-        {
-            try
-            {
-                this.ProcessBlockAddedOrRemoved(item);
+                        try
+                        {
+                            await Task.Delay(2_000, this.cancellation.Token).ConfigureAwait(false);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                        }
+
+                        continue;
+                    }
+
+                    if (nextHeader.Previous.HashBlock != this.IndexerTip.HashBlock)
+                    {
+                        ChainedHeader lastCommonHeader = nextHeader.FindFork(this.IndexerTip);
+
+                        this.logger.LogDebug("Reorg detected. Rewinding till '{0}'.", lastCommonHeader);
+
+                        lock (this.lockObject)
+                        {
+                            foreach (List<AddressBalanceChange> changes in this.addressesIndex.AddressChanges.Values)
+                                changes.RemoveAll(x => x.BalanceChangedHeight > lastCommonHeader.Height);
+                        }
+
+                        this.IndexerTip = lastCommonHeader;
+                        continue;
+                    }
+
+                    // Get next header block and process it.
+                    Block blockToProcess = this.blockStore.GetBlock(nextHeader.HashBlock);
+
+                    if (blockToProcess == null)
+                    {
+                        this.logger.LogDebug("Next block wasn't found. Waiting.");
+
+                        try
+                        {
+                            await Task.Delay(2_000, this.cancellation.Token).ConfigureAwait(false);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                        }
+
+                        continue;
+                    }
+
+                    bool success = this.ProcessBlock(blockToProcess, nextHeader);
+
+                    if (!success)
+                    {
+                        this.logger.LogDebug("Failed to process next block. Waiting.");
+
+                        try
+                        {
+                            await Task.Delay(5_000, this.cancellation.Token).ConfigureAwait(false);
```

Comment:
Seems like this happens when there is a reorg. Could also be a const, we can tweak the values in the future.

Reply:
Done